[BUG]: nb_class.h: new_::wrapper lambda declaration shadows symbol func in capture list, emitting warning

### Problem description

```
my/project/dir/.venv/lib/python3.12/site-packages/nanobind/include/nanobind/nb_class.h:445:25: error: declaration shadows a field of 'new_<type-parameter-0-0, type-parameter-0-1 (type-parameter-0-2...)>' [-Werror,-Wshadow]
  445 |         auto wrapper = [func = (detail::forward_t<Func>) func](handle, Args... args) {
      |                         ^
```

Compiling with Apple clang version 16.0.0 (clang-1600.026.6).

When using -Werror for compilation, the emitted warning breaks builds.

A quick fix could be to lightly rename the captured symbol (e.g. by adding an underscore to the name):

nanobind/include/nanobind/nb_class.h:455:
```
        auto wrapper = [func_ = (detail::forward_t<Func>) func](handle, Args... args) {
            return func_((detail::forward_t<Args>) args...);
        };
```


### Reproducible example code

_No response_

--- a/include/nanobind/nb_class.h
+++ b/include/nanobind/nb_class.h
@@ -442,8 +442,8 @@ struct new_<Func, Return(Args...)> {
               std::is_same_v<detail::intrinsic_t<Args>, kwargs>) + ... + 0);
         detail::wrap_base_new(cl, sizeof...(Args) > num_defaults + num_varargs);
 
-        auto wrapper = [func = (detail::forward_t<Func>) func](handle, Args... args) {
-            return func((detail::forward_t<Args>) args...);
+        auto wrapper = [func_ = (detail::forward_t<Func>) func](handle, Args... args) {
+            return func_((detail::forward_t<Args>) args...);
         };
 
         if constexpr ((std::is_base_of_v<arg, Extra> || ...)) {

